DeeperFunc/Router.swift

```diff
@@ -0,0 +1,206 @@
+        return add(route.map({ intent(flatten($0)) }, { $0.deconstruct(intent).map(parenthesize) }))
+    }
+
+    @discardableResult
+    public func add<A, B, C, D>(_ intent: @escaping (A, B, C, D) -> U, format: String) -> Router {
+        guard let route = format.routePattern else { return self }
+        let typedRoute: RoutePattern<(((A, B), C), D), Path> = route.map({ $0 as? (((A, B), C), D) }, { $0 })
+        return add(intent, route: typedRoute)
+    }
+    
+    @discardableResult
+    public func add<A, B, C, D, E, S: ClosedPatternState>(_ intent: @escaping ((A, B, C, D, E)) -> U, route: RoutePattern<((((A, B), C), D), E), S>) -> Router {
+        return add(route.map({ intent(flatten($0)) }, { $0.deconstruct(intent).map(parenthesize) }))
+    }
+    
+    @discardableResult
+    public func add<A, B, C, D, E>(_ intent: @escaping (A, B, C, D, E) -> U, format: String) -> Router {
+        guard let route = format.routePattern else { return self }
+        let typedRoute: RoutePattern<((((A, B), C), D), E), Path> = route.map({ $0 as? ((((A, B), C), D), E) }, { $0 })
+        return add(intent, route: typedRoute)
+    }
+
+    @discardableResult
+    public func add<A, B, C, D, E, F, S: ClosedPatternState>(_ intent: @escaping ((A, B, C, D, E, F)) -> U, route: RoutePattern<(((((A, B), C), D), E), F), S>) -> Router {
+        return add(route.map({ intent(flatten($0)) }, { $0.deconstruct(intent).map(parenthesize) }))
+    }
+    
+    @discardableResult
+    public func add<A, B, C, D, E, F>(_ intent: @escaping (A, B, C, D, E, F) -> U, format: String) -> Router {
+        guard let route = format.routePattern else { return self }
+        let typedRoute: RoutePattern<(((((A, B), C), D), E), F), Path> = route.map({ $0 as? (((((A, B), C), D), E), F) }, { $0 })
+        return add(intent, route: typedRoute)
+    }
+    
+}
+
+func flatten<A, B, C>(_ t: ((A, B), C)) -> (A, B, C) {
+    return (t.0.0, t.0.1, t.1)
+}
+
+func flatten<A, B, C, D>(_ t: (((A, B), C), D)) -> (A, B, C, D) {
+    return (t.0.0.0, t.0.0.1, t.0.1, t.1)
+}
+
+func flatten<A, B, C, D, E>(_ t: ((((A, B), C), D), E)) -> (A, B, C, D, E) {
+    return (t.0.0.0.0, t.0.0.0.1, t.0.0.1, t.0.1, t.1)
+}
+
+func flatten<A, B, C, D, E, F>(_ t: (((((A, B), C), D), E), F)) -> (A, B, C, D, E, F) {
+    return (t.0.0.0.0.0, t.0.0.0.0.1, t.0.0.0.1, t.0.0.1, t.0.1, t.1)
+}
+
+func parenthesize<A, B, C>(_ t: (A, B, C)) -> ((A, B), C) {
+    return ((t.0, t.1), t.2)
+}
+
+func parenthesize<A, B, C, D>(_ t: (A, B, C, D)) -> (((A, B), C), D) {
+    return (((t.0, t.1), t.2), t.3)
+}
+
+func parenthesize<A, B, C, D, E>(_ t: (A, B, C, D, E)) -> ((((A, B), C), D), E) {
+    return ((((t.0, t.1), t.2), t.3), t.4)
+}
+
+func parenthesize<A, B, C, D, E, F>(_ t: (A, B, C, D, E, F)) -> (((((A, B), C), D), E), F) {
+    return (((((t.0, t.1), t.2), t.3), t.4), t.5)
+}
+
+extension Router {
+
+    public func url(for route: U) -> URL? {
+        return self.route?.print(route).flatMap(url(from:))
```

Comment:
now we can just use composite route to print url and we do it with concrete value of `U`, not a route pattern.